src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/RoutePatternParametersDetector.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.Infrastructure;
+
+internal static class RoutePatternParametersDetector
+{
+    public static ImmutableArray<ISymbol> ResolvedParameters(ISymbol symbol, SemanticModel semanticModel)
+    {
+        var resolvedParameterSymbols = ImmutableArray.CreateBuilder<ISymbol>();
+        var childSymbols = symbol switch
+        {
+            ITypeSymbol typeSymbol => typeSymbol.GetMembers().OfType<IPropertySymbol>().ToImmutableArray().As<ISymbol>(),
+            IMethodSymbol methodSymbol => methodSymbol.Parameters.As<ISymbol>(),
+            _ => throw new InvalidOperationException("Unexpected symbol type: " + symbol)
+        };
+
+        foreach (var child in childSymbols)
+        {
+            if (child.HasAttribute("Microsoft.AspNetCore.Http.AsParametersAttribute", semanticModel))
+            {
+                resolvedParameterSymbols.AddRange(ResolvedParameters(child.GetParameterType(), semanticModel));
+            }
+            else if (HasExplicitNonRouteAttribute(child, semanticModel) || HasSpecialType(child, semanticModel))
```

Comment:
external question: do you want to do these 'skip' checks first?  in case it has both attributes on it?  not sure what desired semantics are on that case.

Reply:
I'll flip them around.